rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,658 @@
+/// 
+/// When this mode is disabled
+#[derive(Resource, Reflect, FromReflect, Debug)]
+struct DevFlyCamera {
+    enabled: bool,
+    /// How fast the camera travels forwards, backwards, left, right, up and down, in world units.
+    movement_speed: Option<f32>,
+    /// How fast the camera turns, in radians per second.
+    turn_speed: Option<f32>,
+}
+
+impl Default for DevFlyCamera {
+    fn default() -> Self {
+        DevFlyCamera {
+            enabled: false,
+            movement_speed: 3.,
+            turn_speed: 10.
+        }
+    }
+}
+
+impl ModalDevTool for DevFlyCamera {
+    fn set_enabled(&mut self, enabled: bool) {
+        self.enabled = enabled;
+    }
+
+    fn is_enabled(&self) -> bool {
+        self.enabled
+    }
+}
+
+impl FromStr for DevFlyCamera {
+    fn from_str(s: &str) -> Result<Self, DevToolParseError>{
+        let parts = s.split_whitespace();
+        let name = parts.iter.next()?;
+        if name != self.name() {
+            return Err(DevToolParseError::InvalidName);
+        }
+
+        let movement_speed: f32 = match parts.iter.next() {
+            Some(string) = parse(string)?,
+            None => Self::default().movement_speed,
+        };
+
+        let turn_speed: f32 = match parts.iter.next() {
+            Some(string) = parse(string)?,
+            None => Self::default().turn_speed,
+        };
+
+        Ok(DevFlyCamera {
+                enabled: true,
+                movmeent_speed,
+                turn_speed,
+            }
+        )
+    }
+}
+```
+
+To configure a dev tool at runtime, simply access the resource, and either mutate or overwrite the `ModalDevTools` struct.
+
+### Dev commands
+
+Dev commands modify the world a single time, and can be called with arguments.
+To model this, we leverage Bevy's existing `Command` trait, which exists to perform complex one-off modifications to the world.
+
+```rust
+/// Dev commands are used by developers to modify the `World` in order to easily debug and test their application.
+/// 
+/// Dev commands can be called with arguments to specify the exact behavior: if you are creating a toolbox, parse the provided arguments 
+/// to construct an instance of the type that implements this type, and then send it as a `Command` to execute it.
+/// 
+/// The documentation on this struct is reflected, and can be read by toolboxes to provide help text to users.
+trait DevCommand: Command + Reflect + FromReflect + FromStr<Err=DevToolParseError> + Debug + 'static {
+    /// The name of this tool, as might be supplied by a command line interface.
+    fn name() -> &'static str {
+        Self::type_name().to_snake_case()
+    }
+
+    /// The metadata for this dev command.
+    fn metadata() -> DevCommandMetadata {
+        DevCommandMetadata {
+            name: self.name(),
+            type_id: Self::type_id(),
+            type_info: Self::type_info(),
```

Comment:
The Self::type_id() and Self::type_info() methods actually require &self as an argument, so cannot be called this way. However, we can get this information via the GetTypeRegistration trait (for example ``` type_info: Self::get_type_registration().type_info() ```) which is automatically implemented via ``` #[derive(Reflect)]```. I suggest to add this trait to the list of parents of ModalDevTool and DevCommand traits